### PR TITLE
Improve some errors

### LINKED ISSUE #81: directly report if no rrule/frule is defined
Right now we hit the fallback definition of `rrule`/frule` which is to  just return `nothing`.
which then gives an error complaining about iteration failing.

We should instead detect that an error ourselves with somehting more clear.
Possibly throwing a `MethodError` which would show what is closest.
(though it would also show the `Any` method that does match but maybe that is ok)

### PR DESCRIPTION
**Pull request by mcabbott**
This tries to make the errors from this package just a hair less inscrutable. 

Edit: closes #81

### CODE REVIEW

File: src/testers.jl [hunk 1]
```diff
@@ -251,7 +251,9 @@ function rrule_test(f, ȳ, xx̄s::Tuple{Any, Any}...; rtol=1e-9, atol=1e-9, fdm
     xs = first.(xx̄s)
     accumulated_x̄ = last.(xx̄s)
     check_inferred && _test_inferred(rrule, f, xs...; fkwargs...)
-    y_ad, pullback = rrule(f, xs...; fkwargs...)
+    res = rrule(f, xs...; fkwargs...)
+    isnothing(res) && throw("no rrule defined for $f, with arguments $(typeof(xs))")
```
Comment (oxinabox) [inline src/testers.jl:255]: What do you think if we throw a `MethodError` hear instead.
Just so it gives the list of near candidates?

Comment (oxinabox) [inline src/testers.jl:255]: Need to use `res===nothing` for 1.0 compatibility

**Review by oxinabox — APPROVED** [review]
LGTM except the think with `isnothing`

File: src/testers.jl (hunk 1 above)
Comment (mcabbott) [reply src/testers.jl:255]: Anything better than `iterate(nothing)` is fine, really.

Comment (mcabbott) [reply src/testers.jl]: New changed.

**Comment by oxinabox** [comment]
Oh sorry I missed this before, but shouldn't we also error if rule is not defined for `frule`?

**Comment by mcabbott** [comment]
Probably, I only fixed things which bit me, so far...

**Comment by mcabbott** [comment]
I've added one for frule too now. (And I think I got the signature right!)

BTW I'm not entirely sure MethodError is the right thing here. It has the right information, but isn't quite the truth -- a method does exist, it's just the fallback one. Maybe an explicit description is clearer?

**Comment by nickrobinson251** [comment]
closes https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/81

**Comment by oxinabox** [comment]
> BTW I'm not entirely sure MethodError is the right thing here. It has the right information, but isn't quite the truth -- a method does exist, it's just the fallback one. Maybe an explicit description is clearer?

Manually thrown MethodErrors do that though.
E.g. https://github.com/JuliaMath/SpecialFunctions.jl/blob/master/src/gamma.jl#L528

**Pull request merged**